homeassistant/components/notify/xmpp.py

```diff
@@ -93,29 +118,217 @@ def __init__(self):
+                _LOGGER.error("Error reading file %s", ex)
+            except FutTimeoutError as ex:
+                _LOGGER.error("The server did not respond in time, %s", ex)
             else:
-                self.send_message(mto=recipient, mbody=message, mtype='chat')
-            self.disconnect(wait=True)
+                _LOGGER.info("Upload success")
+
+                if room:
+                    _LOGGER.info("Sending file to %s", room)
+                    message = self.Message(sto=room, stype='groupchat')
+                else:
+                    _LOGGER.info("Sending file to %s", recipient)
+                    message = self.Message(sto=recipient, stype='chat')
+
+                message['body'] = url
+                # pylint: disable=invalid-sequence-index
+                message['oob']['url'] = url
+                try:
+                    message.send()
+                except (IqError, IqTimeout, XMPPError) as ex:
+                    _LOGGER.error("Could not send image message %s", ex)
+
+        async def upload_file(self, timeout=None):
+            """Upload file to Jabber server and return new URL.
+
+            upload a file with Jabber XEP_0363 from a remote URL or a local
+            file path and return a URL of that file.
+            """
+            if data.get(ATTR_URL_TEMPLATE):
+                _LOGGER.debug(
+                    "Got url template: %s", data.get(ATTR_URL_TEMPLATE))
+                templ = template_helper.Template(
+                    data.get(ATTR_URL_TEMPLATE), hass)
+                get_url = template_helper.render_complex(templ, None)
+                url = await self.upload_file_from_url(
+                    get_url, timeout=timeout)
+            elif data.get(ATTR_URL):
+                url = await self.upload_file_from_url(
+                    data.get(ATTR_URL), timeout=timeout)
+            elif data.get(ATTR_PATH_TEMPLATE):
+                _LOGGER.debug(
+                    "Got path template: %s", data.get(ATTR_PATH_TEMPLATE))
+                templ = template_helper.Template(
+                    data.get(ATTR_PATH_TEMPLATE), hass)
+                get_path = template_helper.render_complex(templ, None)
+                url = await self.upload_file_from_path(
+                    get_path, timeout=timeout)
+            elif data.get(ATTR_PATH):
+                url = await self.upload_file_from_path(
+                    data.get(ATTR_PATH), timeout=timeout)
+            else:
+                url = None
+
+            if url is None:
+                _LOGGER.error("No path or URL found for file")
+
+            return url
+
+        async def upload_file_from_url(self, url, timeout=None):
+            """Upload a file from a URL. Returns a URL.
+
+            uploaded via XEP_0363 and HTTP and returns the resulting URL
+            """
+            _LOGGER.info("Getting file from %s", url)
+
+            def get_url(url):
+                return requests.get(
```

Comment:
Since you are moving to async, it is a bit odd to introduce `requests` usage.

Reply:
yes, I know. Not that familiar with aiohttp yet. requests was my goto. might patch that when I'm up to speed with it.